internal/api/v1/services.go

```diff
@@ -136,6 +140,284 @@ func (sc ServicesController) Index(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	if createRequest.Name == "" {
+		http.Error(w, fmt.Sprintf("Cannot create custom service without a name"),
+			http.StatusBadRequest)
+		return
+	}
+
+	if len(createRequest.Data) < 1 {
+		http.Error(w, fmt.Sprintf("Cannot create custom service without data"),
+			http.StatusBadRequest)
+		return
+	}
+
+	gitea, err := clients.GetGiteaClient()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	exists, err := gitea.OrgExists(org)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	if !exists {
+		http.Error(w, fmt.Sprintf("Organization '%s' does not exist", org),
+			http.StatusNotFound)
+		return
+	}
+
+	cluster, err := kubernetes.GetCluster()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	// Verify that the requested name is not yet used by a different service.
+	_, err = services.Lookup(cluster, org, createRequest.Name)
+	if err == nil {
+		http.Error(w, fmt.Sprintf("Service '%s' already exists", createRequest.Name),
+			http.StatusConflict)
+		return
+	}
+
+	// Create the new service. At last.
+	_, err = services.CreateCustomService(cluster, createRequest.Name, org, createRequest.Data)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	_, err = w.Write([]byte{})
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+}
+
+func (sc ServicesController) Create(w http.ResponseWriter, r *http.Request) {
+	params := httprouter.ParamsFromContext(r.Context())
+	org := params.ByName("org")
+
+	defer r.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	var createRequest models.CatalogCreateRequest
+	err = json.Unmarshal(bodyBytes, &createRequest)
+	if handleError(w, err, http.StatusBadRequest) {
+		return
+	}
+
+	if createRequest.Name == "" {
+		http.Error(w, fmt.Sprintf("Cannot create service without a name"),
+			http.StatusBadRequest)
+		return
+	}
+
+	if createRequest.Class == "" {
+		http.Error(w, fmt.Sprintf("Cannot create service without a service class"),
+			http.StatusBadRequest)
+		return
+	}
+
+	if createRequest.Plan == "" {
+		http.Error(w, fmt.Sprintf("Cannot create service without a service plan"),
+			http.StatusBadRequest)
+		return
+	}
+
+	gitea, err := clients.GetGiteaClient()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	exists, err := gitea.OrgExists(org)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	if !exists {
+		http.Error(w, fmt.Sprintf("Organization '%s' does not exist", org),
+			http.StatusNotFound)
+		return
+	}
+
+	cluster, err := kubernetes.GetCluster()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	// Verify that the requested name is not yet used by a different service.
+	_, err = services.Lookup(cluster, org, createRequest.Name)
+	if err == nil {
```

Comment:
same as above, need to check what the error is

Reply:
Checking...